inject/src/main/java/io/micronaut/inject/annotation/EvaluatedConvertibleValuesMap.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright 2017-2022 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.inject.annotation;
+
+import io.micronaut.context.BeanContext;
+import io.micronaut.context.BeanDefinitionAware;
+import io.micronaut.context.ContextConfigurable;
+import io.micronaut.core.annotation.Internal;
+import io.micronaut.core.annotation.Nullable;
+import io.micronaut.core.convert.ArgumentConversionContext;
+import io.micronaut.core.convert.ConversionService;
+import io.micronaut.core.convert.value.ConvertibleValues;
+import io.micronaut.core.convert.value.ConvertibleValuesMap;
+import io.micronaut.core.expression.EvaluatedExpression;
+import io.micronaut.inject.BeanDefinition;
+
+import java.util.Collection;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Version of {@link ConvertibleValuesMap} that is aware of evaluated expressions.
+ *
+ * @param <V> The generic value
+ */
+@Internal
+public class EvaluatedConvertibleValuesMap<V> implements ConvertibleValues<V>
+{
+    private final BeanContext beanContext;
+    private final BeanDefinition<?> owningBean;
+    private final Object[] args;
+    private final ConvertibleValues<V> delegateValues;
+
+    EvaluatedConvertibleValuesMap(@Nullable BeanContext beanContext,
+                                  @Nullable BeanDefinition<?> owningBean,
+                                  Object[] args,
+                                  ConvertibleValues<V> delegateValues) {
+        this.beanContext = beanContext;
+        this.owningBean = owningBean;
+        this.args = args;
+        this.delegateValues = delegateValues;
+    }
+
+    @Override
+    public Set<String> names() {
+        return delegateValues.names();
+    }
+
+    @Override
+    public <T> Optional<T> get(CharSequence name,
+                               ArgumentConversionContext<T> conversionContext) {
+        V value = delegateValues.getValue(name);
+        if (value instanceof EvaluatedExpression expression) {
```

Comment:
This should probably reuse some mapping function `Function<Object, Object>` instead of duplicating the logic.